### ISSUE TITLE
False positive `no-self-use` when a method defines a sub-class that itself does not use `self` in a method

### ISSUE THREAD

**Issue by danielrozenberg** [pylint-dev/pylint#3705]
### Steps to reproduce
1. copy the code from below into a Python file and run pylint on it :)

### Current behavior

The following file generates a report with a false positive for `no-self-use`:

false_positive.py
```python
"""Test file."""
# pylint: disable=missing-class-docstring,missing-function-docstring,too-few-public-methods
import unittest
from class_generator import ExternalClassGenerator

class MyTest(unittest.TestCase):

    def test_one(self, value):

        with self.assertRaises(TypeError):
            class _MySubClass(ExternalClassGenerator(value)):

                def my_sub_method(self):
                    return value
```

Report:
```log
************* Module false_positive
false_positive.py:13:16: R0201: Method could be a function (no-self-use)
false_positive.py:8:4: R0201: Method could be a function (no-self-use)
```

The `self` for the method `test_one` on line 8 is used in line 10
The `self` for the method `my_sub_method` (inside the sub-class) on line 13 is indeed a correct `no-self-use`

**Both** errors disappear if I do one of the following:
* Change the inner class to just:
  ```python
            class _MySubClass(ExternalClassGenerator(value)):
                pass
  ```
* Use `self` in the `my_sub_method` method

### Expected behavior

Only `my_sub_method` should cause a `no-self-use`

### pylint --version output
```
pylint 2.5.2
astroid 2.4.1
Python 3.8.2 (default, Apr 27 2020, 15:53:34) 
[GCC 9.3.0]
```

**Comment by clavedeluna** [comment]
Trying to reproduce it but the original issue does not show up. Here's code with `from class_generator import ExternalClassGenerator` removed since idk what that is

```
"""Test file."""
# pylint: disable=missing-class-docstring,missing-function-docstring,too-few-public-methods
import unittest

class ExternalClassGenerator:

    def __init__(self, value):
        self.value = value

class MyTest(unittest.TestCase):

    def test_one(self, value):

        with self.assertRaises(TypeError):
            class _MySubClass(ExternalClassGenerator(value)):

                def my_sub_method(self):
                    return value
```

results in

```
test.py:15:12: E0239: Inheriting 'ExternalClassGenerator(value)', which is not a class. (inherit-non-class)
test.py:15:12: W0612: Unused variable '_MySubClass' (unused-variable)
```

so original issue of false positive on `test_one` no longer shows up (good!) but the previously reported `no-self-use` on `my_sub_method` no longer shows up. Though that actually seems fine to me. a method doesn't have to always use self. 

Simplifying the original code to 
```
class _MySubClass:

    def my_sub_method(self):
        return value
```

there's really no reason for `my_sub_method` to ever raise `no-self-use`

So I think this issue is OK to close.

**Comment by Pierre-Sassoulas** [comment]
This time I think I reproduced this issue june 30. ``no-self-use`` is now an extension did you load it with ``--load-plugins`` ?

**Comment by Pierre-Sassoulas** [comment]
I generally copy paste the example in a ``a.py`` file then ``pylint a.py --enable=all --enable-all-extensions`` when trying to reproduce :)
